src/components/containers/TraceAccordion.js

```diff
@@ -10,38 +10,109 @@ import {Tab, Tabs, TabList, TabPanel} from 'react-tabs';
-    const filteredData = data.filter(t => {
-      if (excludeFits) {
+    const base = props.canGroup ? context.fullData : context.data;
+
+    this.filteredTracesFullDataPositions = [];
+    this.filteredTraces = base.filter((t, i) => {
+      if (props.excludeFits) {
         return !(t.transforms && t.transforms.every(tr => tr.type === 'fit'));
       }
+      this.filteredTracesFullDataPositions.push(i);
       return true;
     });
+  }
 
-    const individualTraces =
-      filteredData.length &&
-      filteredData.map((d, i) => {
-        return (
-          <TraceFold
-            key={i}
-            traceIndexes={[i]}
-            canDelete={canAdd}
-            messageIfEmpty={messageIfEmptyFold}
-          >
-            {children}
-          </TraceFold>
-        );
-      });
+  renderGroupedTraceFolds() {
+    if (!this.filteredTraces.length || this.filteredTraces.length < 2) {
+      return null;
+    }
+
+    const dataArrayPositionsByTraceType = {};
+    const fullDataArrayPositionsByTraceType = {};
+
+    this.filteredTraces.forEach((trace, index) => {
+      const traceType = plotlyTraceToCustomTrace(trace.type);
+      if (!dataArrayPositionsByTraceType[traceType]) {
+        dataArrayPositionsByTraceType[traceType] = [];
+      }
+
+      if (!fullDataArrayPositionsByTraceType[traceType]) {
+        fullDataArrayPositionsByTraceType[traceType] = [];
+      }
+
+      dataArrayPositionsByTraceType[traceType].push(trace.index);
+      fullDataArrayPositionsByTraceType[traceType].push(
+        this.filteredTracesFullDataPositions[index]
+      );
+    });
+
+    return Object.keys(fullDataArrayPositionsByTraceType).map((type, index) => {
+      return (
+        <TraceFold
+          key={index}
+          traceIndexes={dataArrayPositionsByTraceType[type]}
+          name={type}
+          fullDataArrayPosition={fullDataArrayPositionsByTraceType[type]}
+        >
+          {this.props.children}
+        </TraceFold>
+      );
+    });
+  }
+
+  renderUngroupedTraceFolds() {
+    if (!this.filteredTraces.length) {
+      return null;
+    }
+
+    return this.filteredTraces.map((d, i) => {
+      return (
+        <TraceFold
+          key={i}
+          traceIndexes={[d.index]}
+          canDelete={this.props.canAdd}
+          messageIfEmpty={this.props.messageIfEmptyFold}
+          fullDataArrayPosition={[this.filteredTracesFullDataPositions[i]]}
+        >
+          {this.props.children}
+        </TraceFold>
+      );
+    });
+  }
+
+  renderTraceFolds() {
+    if (!this.filteredTraces.length) {
+      return null;
+    }
+
+    return this.filteredTraces.map((d, i) => {
+      return (
+        <TraceFold
+          key={i}
+          traceIndexes={[i]}
+          canDelete={this.props.canAdd}
+          messageIfEmpty={this.props.messageIfEmptyFold}
+        >
+          {this.props.children}
+        </TraceFold>
+      );
+    });
+  }
+
+  render() {
```

Comment:
Nice refactoring. This is a pretty hairy component and it's already looking a lot better! At some point I wonder if we can't break it apart more completely by having a `GroupedTraceAccordion` but not today.